File: src/controllers/auth/passwordReset.controllers.js
import * as passwordResetService from "../../services/auth/passwordReset.service.js"

/**
 * @param {import("express").Request} req
 * @param {import("express").Response} res
 */
export const requestPasswordReset = async (req, res) => {
  const { email } = req.body
  try {
    const resp = await passwordResetService.requestPasswordReset(email)

    if (resp.error) return res.status(400).send(resp.error)

    req.session.passwordReset = {
      email,
      emailConfirmed: false,
      passwordResetToken: resp.passwordResetToken,
      passwordResetTokenExpires: resp.passwordResetTokenExpires,
    }

    req.sessionOptions.maxAge = 60 * 60 * 1000
    req.sessionOptions.path = "/api/auth/forgot_password/confirm_email"

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const confirmEmail = async (req, res) => {
  const { token: inputToken } = req.body

  try {
    if (!req.session?.passwordReset) return res.sendStatus(401)

    const passwordResetSessionData = req.session.passwordReset

    const resp = passwordResetService.confirmEmail({
      inputToken,
      ...passwordResetSessionData,
    })

    if (resp.error) return res.status(400).send(resp.error)

    req.session.passwordReset = {
      email: passwordResetSessionData.email,
    }

    req.sessionOptions.maxAge = 60 * 60 * 1000
    req.sessionOptions.path = "/api/auth/forgot_password/reset_password"

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const resetPassword = async (req, res) => {
  try {
    if (!req.session?.passwordReset) return res.sendStatus(401)

    const { email } = req.session.passwordReset

    const { newPassword } = req.body

    const resp = await passwordResetService.resetPassword(email, newPassword)

    req.session = null

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

File: src/controllers/auth/passwordResetControllers/passwordResetControllers.go
package passwordResetControllers

import (
	"i9lyfe/src/helpers"
	"i9lyfe/src/services/auth/passwordResetService"
	"time"

	"github.com/gofiber/fiber/v3"
	"github.com/vmihailenco/msgpack/v5"
)

func RequestPasswordReset(c fiber.Ctx) error {
	ctx := c.Context()

	var body requestPasswordResetBody

	err := c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err := body.Validate(); err != nil {
		return err
	}

	respData, sessionData, err := passwordResetService.RequestPasswordReset(ctx, body.Email)
	if err != nil {
		return err
	}

	reqSession := map[string]any{
		"passwordReset": sessionData,
	}

	c.Cookie(helpers.Session(reqSession, "/api/auth/forgot_password/confirm_email", int(time.Hour/time.Second)))

	return c.MsgPack(respData)
}

func ConfirmEmail(c fiber.Ctx) error {
	ctx := c.Context()

	sessionData := c.Locals("passwordReset_sess_data").(msgpack.RawMessage)

	var body confirmEmailBody

	err := c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err := body.Validate(); err != nil {
		return err
	}

	respData, newSessionData, err := passwordResetService.ConfirmEmail(ctx, sessionData, body.Token)
	if err != nil {
		return err
	}

	reqSession := map[string]any{
		"passwordReset": newSessionData,
	}

	c.Cookie(helpers.Session(reqSession, "/api/auth/forgot_password/reset_password", int(time.Hour/time.Second)))

	return c.MsgPack(respData)
}

func ResetPassword(c fiber.Ctx) error {
	ctx := c.Context()

	sessionData := c.Locals("passwordReset_sess_data").(msgpack.RawMessage)

	var body resetPasswordBody

	err := c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err := body.Validate(); err != nil {
		helpers.LogError(err)
		return err
	}

	respData, err := passwordResetService.ResetPassword(ctx, sessionData, body.NewPassword)
	if err != nil {
		return err
	}

	c.ClearCookie()

	return c.MsgPack(respData)
}
